### ISSUE TITLE
Question about the number of voxels is less than one

### ISSUE THREAD

**Issue by uJunLI** [OpenRobotLab/AnySplat#64]
![Image](https://github.com/user-attachments/assets/23ac1461-ac7a-4c17-a097-59b281444395)
I encounter this assertion error, but the model outputs dense point map, so what is the reason for this happening

**Comment by uJunLI** [comment]
The strange thing is that when I trained with four GPUs, the error occurred simultaneously on all of them, so I suspect it might not be a runtime issue.

**Comment by AntaresAlice** [comment]
I’ve also encountered this problem. When I checked other elements like scene_scale, they turned out to be NaN. I suspect the issue is related to the training dataset, since I noticed there were too many black images in my original dataset. After switching my training dataset to dl3dv, it runs well now.

**Comment by uJunLI** [comment]
> I’ve also encountered this problem. When I checked other elements like scene_scale, they turned out to be NaN. I suspect the issue is related to the training dataset, since I noticed there were too many black images in my original dataset. After switching my training dataset to dl3dv, it runs well now.

Thanks for your share I suspect that the issue is related to gradient explosion. This is because the problem only emerged after I made modifications to the code, while the training dataset has consistently been DL3DV. Since the data remained unchanged, the abnormal behavior is more likely caused by parameter initialization or architectural adjustments leading to numerical instability. However, I still cannot be certain, as it is unclear why reasonable code modifications would directly trigger gradient explosion.

**Comment by jianglh-WHU** [comment]
You can print more informations to debug, like the attributes of Gaussian primitives, scene_scale and etc.

**Comment by ricshaw** [comment]
also encountering this error while training on DL3DV dataset

**Comment by ricshaw** [comment]
anchor_feats is all nans

**Comment by jianglh-WHU** [comment]
What's your training setting? Do you use my training script without any modification?

**Comment by ricshaw** [comment]
I am training with DL3DV. I have made minor modifications, e.g. to data loading paths, batch size=1, num_context_views=8, max_img_per_gpu=8, and my GPU doesn't bfloat16. Just trying to debug it now

**Comment by ricshaw** [comment]
For me, the nans are appearing after calling the aggregator, as the aggregated_tokens_list contains nans. So I believe the gradients are exploding....

**Comment by ricshaw** [comment]
I still can't work out why this is happening. Suddenly at training step 541 I get nans in the aggregated_tokens_list:

```
dtype_ = torch.bfloat16 if torch.cuda.get_device_capability()[0] >= 8 else torch.float16
with torch.amp.autocast("cuda", enabled=True, dtype=dtype_):
    # image_has_nan = torch.isnan(image).any()
    # print("image has nan:", image_has_nan, image.shape, image.min(), image.max(), image.dtype)
    aggregated_tokens_list, patch_start_idx = self.aggregator(
        image.to(torch.bfloat16),
        intermediate_layer_idx=self.cfg.intermediate_layer_idx,
    )
```

**Comment by ricshaw** [comment]
I managed to fix my nan problem. I think it was because I was not using a GPU that supported bfloat16, so I had to change to torch.float16, but I was mixing precision modes. I changed my yaml to precision: "16-mixed", but in the code the model was still being cast to bfloat16, resulting in no loss or gradient scaling. Removing the model cast made it work.

**Comment by JamChaos** [comment]
@uJunLI  Did you finally solve it?
